src/udev/udevadm-hwdb.c

```diff
@@ -35,11 +35,69 @@
  * Uses a Patricia/radix trie to index all matches for efficient lookup.
  */
 
-static const char * const conf_file_dirs[] = {
-        UDEV_HWDB_DIR,
-        UDEV_LIBEXEC_DIR "/hwdb.d",
-        NULL
-};
+static ssize_t follow_path (const char *path_variable, char **destination, size_t max) {
+        size_t n = 0;
+        if (path_variable != NULL) {
+                while (path_variable[0] == ':') {
+                      path_variable++;
+                }
+                while (path_variable[0] != '\0') {
+                        const char *end = strchr(path_variable, ':');
+                        if (end == NULL) {
+                                end = path_variable + strlen(path_variable);
+                        }
+                        if (n < max) {
+                                destination[n] = strndup(path_variable, end - path_variable);
+                                if (destination[n] == NULL) {
+                                        return -ENOMEM;
+                                }
+                        }
+                        n++;
+                        path_variable = end;
+                        while (path_variable[0] == ':') {
+                                path_variable++;
+                        }
+                }
+        }
+        if (n < max) {
+                destination[n] = NULL;
+        }
+        return n;
+}
+
+static char ** list_conf_file_dirs (void) {
+        static const char *main_hwdb_dir = UDEV_HWDB_DIR;
+        static const char *libexec_dir = UDEV_LIBEXEC_DIR "/hwdb.d";
+        const char *path_variable = getenv ("UDEV_HWDB_PATH");
+        ssize_t path_length = 0;
+        if (path_variable != NULL) {
+                path_length = follow_path(path_variable, NULL, 0);
+                if (path_length < 0) {
+                      return NULL;
+                }
+        }
+        char **all_dirs = malloc((path_length + 2 + 1) * sizeof (char *));
+        if (all_dirs == NULL) {
+                return NULL;
+        }
+        for (ssize_t i = 0; i < path_length + 2 + 1; i++) {
+                all_dirs[i] = NULL;
+        }
+        ssize_t error = follow_path(path_variable, all_dirs, path_length + 1);
+        all_dirs[path_length] = strdup(main_hwdb_dir);
+        all_dirs[path_length + 1] = strdup(libexec_dir);
+        all_dirs[path_length + 2] = NULL;
+        if (error < 0
+            || all_dirs[path_length] == NULL
+            || all_dirs[path_length + 1] == NULL) {
+                for (ssize_t i = 0; i < path_length + 2 + 1; i++) {
+                        free(all_dirs[i]);
+                }
```

Comment:
`free(NULL)` is harmless in many environments, but isn't it better to check `if (all_dirs[i])` before the free?